src/rewriter.coffee

```diff
@@ -548,20 +548,86 @@ exports.Rewriter = class Rewriter
   # `OUTDENT` tokens should always be positioned at the last character of the
   # previous token, so that AST nodes ending in an `OUTDENT` token end up with a
   # location corresponding to the last “real” token under the node.
-  fixOutdentLocationData: ->
+  fixIndentationLocationData: ->
+    @allComments ?= extractAllCommentTokens @tokens
+    findPrecedingComment = (token, {afterPosition, indentSize, first, indented}) =>
+      tokenStart = token[2].range[0]
+      matches = (comment) ->
+        if comment.outdented
+          return no unless indentSize? and comment.indentSize > indentSize
+        return no if indented and not comment.indented
+        return no unless comment.locationData.range[0] < tokenStart
+        return no unless comment.locationData.range[0] > afterPosition
+        yes
+      if first
+        lastMatching = null
+        for comment in @allComments by -1
+          if matches comment
+            lastMatching = comment
+          else if lastMatching
+            return lastMatching
+        return lastMatching
+      for comment in @allComments when matches comment by -1
```

Comment:
There is also an optimization opportunity here when finding the closest preceding comment

Again I'd rather tackle optimization as a separate PR after this if that works for you